ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractLoader.scala

```diff
@@ -0,0 +1,155 @@
+      (KEY, LoggingContext, Promise[Option[VALUE]])
+    ]],
+    maxBatchSize: Int,
+    parallelism: Int,
+)(implicit executionContext: ExecutionContext, materializer: Materializer)
+    extends Loader[KEY, VALUE] {
+  private implicit val contextualizedErrorLogger: ContextualizedErrorLogger =
+    DamlContextualizedErrorLogger.forClass(this.getClass)
+
+  private val (queue, done) = createQueue() // TODO maybe set up autorestart?
+    .via(
+      BatchN(
+        maxBatchSize = maxBatchSize,
+        maxBatchCount = parallelism,
+      )
+    )
+    .mapAsyncUnordered(parallelism) { batch =>
+      batchLoad(
+        batch.view.map { case (key, loggingContext, _) => key -> loggingContext }.toSeq
+      ).transform {
+        case Success(resultMap) =>
+          batch.view.foreach { case (key, _, promise) =>
+            promise.success(resultMap.get(key))
+            ()
+          }
+          Success(())
+
+        case Failure(t) =>
+          batch.view.foreach { case (_, _, promise) =>
+            promise.failure(
+              LedgerApiErrors.InternalError
+                .Generic("Contract lookup failed", Some(t))
+                .asGrpcError
+            )
+            ()
+          }
+          Success(())
+      }
+    }
+    .toMat(Sink.ignore)(Keep.both)
+    .run()
+
+  override def load(key: KEY)(implicit loggingContext: LoggingContext): Future[Option[VALUE]] = {
+    val promise = Promise[Option[VALUE]]()
+    queue.offer((key, loggingContext, promise)) match {
+      case QueueOfferResult.Enqueued => promise.future
+
+      case QueueOfferResult.Dropped =>
+        Future.failed(
+          LedgerApiErrors.ParticipantBackpressure
+            .Rejection("Too many pending contract lookups")
+            .asGrpcError
+        )
+
+      // these should never happen, if this service is closed in the right order
+      case QueueOfferResult.QueueClosed => Future.failed(new IllegalStateException("Queue closed"))
+      case QueueOfferResult.Failure(t) => Future.failed(new IllegalStateException(t.getMessage))
+    }
+  }
+
+  def closeAsync(): Future[Unit] = {
+    queue.complete()
+    done.map(_ => ())
+  }
+}
+
+trait ContractLoader extends Loader[(ContractId, Offset), ContractStorageBackend.RawContractState]
+
+object ContractLoader {
+  def create(
+      contractStorageBackend: ContractStorageBackend,
+      dbDispatcher: DbDispatcher,
+      metrics: Metrics,
+      maxQueueSize: Int,
+      maxBatchSize: Int,
+      parallelism: Int,
+  )(implicit
+      materializer: Materializer,
+      executionContext: ExecutionContext,
+  ): ResourceOwner[ContractLoader] = {
+    ResourceOwner
+      .forReleasable(() =>
+        new AkkaStreamParallelBatchedLoader[
+          (ContractId, Offset),
+          ContractStorageBackend.RawContractState,
+        ](
+          batchLoad = { batch =>
+            val ((_, latestValidAtOffset), usedLoggingContext) = batch.maxBy(_._1._2)
```

Comment:
@tudor-da do you think it is okay to simply take the latest here?

Reply:
As discussed online, I think it's fine!

Reply:
thanks for taking the time to wrap your head around this